src/gifcodec.c

```diff
@@ -198,40 +190,46 @@ DGifSlurpMono(GifFileType * GifFile, SavedImage *TrailingExtensions)
 
 		switch (RecordType) {
 			case IMAGE_DESC_RECORD_TYPE: {
-				/* This call will leak GifFile->Image.ColorMap; there's a fixme in the DGifGetImageDesc code */
 				if (DGifGetImageDesc(GifFile) == GIF_ERROR) {
 					return (GIF_ERROR);
 				}
 
 				sp = &GifFile->SavedImages[GifFile->ImageCount - 1];
+				/* Allocate memory for the image */
+				if (sp->ImageDesc.Width < 0 && sp->ImageDesc.Height < 0 &&
+						sp->ImageDesc.Width > (INT_MAX / sp->ImageDesc.Height)) {
+					return GIF_ERROR;
+				}
+
 				ImageSize = sp->ImageDesc.Width * sp->ImageDesc.Height;
-				if (ImageSize == 0)
+				if (ImageSize > (SIZE_MAX / sizeof(GifPixelType))) {
 					return GIF_ERROR;
+				}
 
-				sp->RasterBits = (BYTE*) GdipAlloc (ImageSize * sizeof (GifPixelType));
+				sp->RasterBits = (unsigned char *) malloc(ImageSize * sizeof(GifPixelType));
 				if (sp->RasterBits == NULL) {
 					return GIF_ERROR;
 				}
 
-				if (GifFile->Image.Interlace) {
-					/* first start at line 0 and read every 8th lines */
-					if (gif_read_interlace (GifFile, sp, 0, 8) == GIF_ERROR)
-						return GIF_ERROR;
-					/* then start at line 4 and read every 8th lines */
-					if (gif_read_interlace (GifFile, sp, 4, 8) == GIF_ERROR)
-						return GIF_ERROR;
-					/* then start at line 2 and read every 4th lines */
-					if (gif_read_interlace (GifFile, sp, 2, 4) == GIF_ERROR)
-						return GIF_ERROR;
-					/* then start at line 1 and read every 2th lines */
-					if (gif_read_interlace (GifFile, sp, 1, 2) == GIF_ERROR)
-						return GIF_ERROR;
-					/* all lines are read */
-				} else {
-					if (DGifGetLine(GifFile, sp->RasterBits, ImageSize) == GIF_ERROR) {
-						return (GIF_ERROR);
+				if (sp->ImageDesc.Interlace) {
+					int i, j;
+					/* 
+					* The way an interlaced image should be read - 
+					* offsets and jumps...
+					*/
+					int InterlacedOffset[] = { 0, 4, 2, 1 };
+					int InterlacedJumps[] = { 8, 8, 4, 2 };
+					/* Need to perform 4 passes on the image */
+					for (i = 0; i < 4; i++) {
+						for (j = InterlacedOffset[i]; j < sp->ImageDesc.Height; j += InterlacedJumps[i]) {
+							if (DGifGetLine(GifFile, sp->RasterBits + j * sp->ImageDesc.Width, sp->ImageDesc.Width) == GIF_ERROR)
+								return GIF_ERROR;
+						}
```

Comment:
This is essentially the same code as before except for the removed `* sizeof (GifPixelType)`. I found the previous code slightly more readable, but the change seems to be correct.

Reply:
I don't think we should be writing our own code here though - I'd prefer to match upstream giflib rather than roll our own. Although it probably doesn't matter here - I have no strong opinion

Reply:
No strong opinion either. I didn't realize that it was copied upstream code in the first place. Just wanted to point out what actually changed and could have caused the memory corruption in the first place.